hydroflow/tests/surface_syntax.rs

```diff
@@ -0,0 +1,45 @@
+use hydroflow::hydroflow_parser;
+
+#[test]
+pub fn test_parser_basic() {
+    // hydroflow_parser! {
+    //     edges_input = (input() ->);
+
+    //     init_vertex = (seed([0]) ->);
+    //     // loop_vertices = (->);
+    //     out_vertices = (-> for_each(|x| println!("Reached: {}", x)));
+
+    //     reached_vertices = (merge[init_vertex, loop_vertices] -> map(|v| (v, ())));
+
+    //     (join[reached_vertices, edges_input] -> map(|(_src, ((), dst))| dst) -> dedup() -> tee[loop_vertices, out_vertices]);
+
+    //     // x = (a -> b() -> c() -> (a -> b -> c) -> p);
+    //     // b = (a -> b() -> c() -> (a -> b -> c) -> p);
+    //     // x = (a -> b() -> c() -> (a -> b -> c) -> p);
+    //     // x = (a -> b() -> c() -> (a -> b -> c) -> p);
+    // }
+
+    hydroflow_parser! {
+        edges_input = input();
```

Comment:
So as I understand it, each variable definition is effectively a subgraph? I wonder if it would be easier to understand if the inputs / outputs were explicitly declared (making each variable effectively into a function). This is a _lot_ more work though, so may not be worth it.

I'm imagining something like this:
```rust
edges_input[() -> (out)] = input() -> out;

...

reached_vertices[(init, loop) -> reached] = ((init, loop) -> merge() -> map(|v| (v, ()));
```

Reply:
Yes that they are a small piece of the graph. But no in the in-out trees scheduled-layer sense.

In the future we may want to have some sort of encapsulation/reusability for assembling operators together. But for now want to keep things simple

Reply:
Makes sense. I think as an iteration in the future it'll generally be nice to explore how we can have richer naming on inputs and outputs. I think the indexing is fine for now since subgraphs don't have that many inputs or outputs but if we want to support richer composition it'll be nice for users to be able to give names.